File: app/src/main/java/za/org/grassroot2/view/fragment/SingleTextInputFragment.java
package za.org.grassroot2.view.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import org.greenrobot.eventbus.EventBus;

import butterknife.BindView;
import io.reactivex.Observable;
import za.org.grassroot2.R;
import za.org.grassroot2.dagger.activity.ActivityComponent;
import za.org.grassroot2.rxbinding.RxTextView;
import za.org.grassroot2.rxbinding.RxView;
import za.org.grassroot2.rxbinding.RxViewUtils;
import za.org.grassroot2.view.SingleInputNextOtherView;

public class SingleTextInputFragment extends TextInputFragment implements SingleInputNextOtherView {

    private static final String INPUT_LABEL_RES = "INPUT_LABEL_RES";
    private static final String INPUT_HINT_RES = "INPUT_HINT_RES";
    private static final String NEXT_BUTTON_RES = "NEXT_BUTTON_RES";

    private int nextBtnRes;
    private int inputLabelRes;
    private int inputHintRes;

    @BindView(R.id.explanation_text) TextView explanation;
    @BindView(R.id.button_back)
    ImageButton backButton;
    @BindView(R.id.button_next) Button nextButton;
    @BindView(R.id.inputLabel)
    TextView inputLabel;
    @BindView(R.id.text_input_field)
    EditText inputField;

    public SingleTextInputFragment() {
    }

    public static SingleTextInputFragment newInstance(int headerTextRes, int explanTextRes, int inputLabelRes, int inputHintRes, int nextButtonRes) {
        SingleTextInputFragment fragment = new SingleTextInputFragment();
        Bundle args = new Bundle();
        addStandardArgs(args, headerTextRes, explanTextRes);
        args.putInt(INPUT_LABEL_RES, inputLabelRes);
        args.putInt(INPUT_HINT_RES, inputHintRes);
        args.putInt(NEXT_BUTTON_RES, nextButtonRes);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            Bundle args = getArguments();
            nextBtnRes = args.getInt(NEXT_BUTTON_RES);
            inputLabelRes = args.getInt(INPUT_LABEL_RES);
            inputHintRes = args.getInt(INPUT_HINT_RES);
        }
    }

    @Override
    protected void onInject(ActivityComponent activityComponent) {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = super.onCreateView(inflater, container, savedInstanceState);
        header.setText(headerTextRes);
        explanation.setText(explanTextRes);

        nextButton.setText(nextBtnRes);
        inputLabel.setText(inputLabelRes);
        inputField.setHint(inputHintRes);

        lifecyclePublisher.onNext(ACTION_FRAGMENT_VIEW_CREATED);
        return v;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        Observable<CharSequence> editTextNext = RxTextView
                .editorActions(inputText, RxViewUtils.imeNextDonePredicate())
                .filter(integer -> {
                    return inputText != null; // because Android
                })
                .map(integer -> inputText.getText());
        Observable<CharSequence> nextButtonClicked = RxView
                .clicks(nextButton).map(o -> inputText.getText());

        Observable<CharSequence> backClick = RxView.clicks(backButton)
                .map(o -> inputText.getText());

        disposables.add(RxTextView.textChanges(inputText).subscribe(charSequence -> {
            if (charSequence.length()>1) {
                toggleNextDoneButton(true);
            }
        }, Throwable::printStackTrace));

        disposables.add(Observable.merge(editTextNext, nextButtonClicked).subscribe(charSequence ->
                EventBus.getDefault().post(new SingleInputTextEvent(this, SingleInputTextEventType.DONE, charSequence)), Throwable::printStackTrace));

        disposables.add(backClick.subscribe(charSequence ->
                EventBus.getDefault().post(new SingleInputTextEvent(this, SingleInputTextEventType.BACK, "")), Throwable::printStackTrace));
    }

    @Override
    public int getLayoutResourceId() {
        return R.layout.fragment_single_text_input;
    }

    @Override
    public void toggleNextDoneButton(boolean enabled) {
        nextButton.setEnabled(enabled);
    }

    @Override
    public void toggleBackOtherButton(boolean enabled) {
        backButton.setEnabled(enabled);
        backButton.setVisibility(enabled ? View.VISIBLE : View.INVISIBLE);
    }


    public enum SingleInputTextEventType {CHANGE, DONE, BACK}

    public static class SingleInputTextEvent {
        public final GrassrootFragment source;
        public final SingleInputTextEventType type;
        public final CharSequence value;

        public SingleInputTextEvent(GrassrootFragment source, SingleInputTextEventType type, CharSequence value) {
            this.source = source;
            this.type = type;
            this.value = value;
        }
    }
}

File: app/src/main/java/za/org/grassroot2/view/fragment/SingleTextInputFragment.kt
package za.org.grassroot2.view.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.jakewharton.rxbinding2.view.RxView
import com.jakewharton.rxbinding2.widget.RxTextView
import io.reactivex.Observable
import kotlinx.android.synthetic.main.fragment_single_text_input.*
import org.greenrobot.eventbus.EventBus
import timber.log.Timber
import za.org.grassroot2.R
import za.org.grassroot2.dagger.activity.ActivityComponent
import za.org.grassroot2.rxbinding.RxViewUtils
import za.org.grassroot2.view.SingleInputNextOtherView

class SingleTextInputFragment : TextInputFragment(), SingleInputNextOtherView {
    override val layoutResourceId: Int
        get() = R.layout.fragment_single_text_input

    private var nextBtnRes: Int = 0
    private var inputLabelRes: Int = 0
    private var inputHintRes: Int = 0

    override fun getInputText(): TextView = inputText

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            val args = arguments
            nextBtnRes = args!!.getInt(NEXT_BUTTON_RES)
            inputLabelRes = args.getInt(INPUT_LABEL_RES)
            inputHintRes = args.getInt(INPUT_HINT_RES)
        }
    }

    override fun onInject(activityComponent: ActivityComponent) {}

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = super.onCreateView(inflater, container, savedInstanceState)
        header.setText(headerTextRes)
        explanation!!.setText(explanTextRes)

        nextButton.setText(nextBtnRes)
        inputLabel.setText(inputLabelRes)
        inputText.setHint(inputHintRes)

        lifecyclePublisher.onNext(GrassrootFragment.ACTION_FRAGMENT_VIEW_CREATED)
        return v
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val editTextNext = RxTextView
                .editorActions(inputText, RxViewUtils.imeNextDonePredicate())
                .filter { integer ->
                    inputText != null // because Android
                }
                .map<CharSequence> { integer -> inputText.text }
        val nextButtonClicked = RxView
                .clicks(nextButton!!).map<CharSequence> { o -> inputText.text }

        val backClick = RxView.clicks(backButton!!)
                .map<CharSequence> { o -> inputText.text }

        disposables.add(RxTextView.textChanges(inputText).subscribe({ charSequence ->
            if (charSequence.length > 1) {
                toggleNextDoneButton(true)
            } }, { Timber.e(it) }))

        disposables.add(Observable.merge(editTextNext, nextButtonClicked)
                .subscribe({ charSequence ->
                    EventBus.getDefault().post(SingleInputTextEvent(this, SingleInputTextEventType.DONE, charSequence))
                }, { it.printStackTrace() }))

        disposables.add(backClick.subscribe({ charSequence ->
                    EventBus.getDefault().post(SingleInputTextEvent(this, SingleInputTextEventType.BACK, "")) }, { Timber.e(it) }))
    }

    override fun toggleNextDoneButton(enabled: Boolean) {
        nextButton!!.isEnabled = enabled
    }

    override fun toggleBackOtherButton(enabled: Boolean) {
        backButton!!.isEnabled = enabled
        backButton!!.visibility = if (enabled) View.VISIBLE else View.INVISIBLE
    }


    enum class SingleInputTextEventType {
        CHANGE, DONE, BACK
    }

    class SingleInputTextEvent(val source: GrassrootFragment, val type: SingleInputTextEventType, val value: CharSequence)

    companion object {

        private val INPUT_LABEL_RES = "INPUT_LABEL_RES"
        private val INPUT_HINT_RES = "INPUT_HINT_RES"
        private val NEXT_BUTTON_RES = "NEXT_BUTTON_RES"

        fun newInstance(headerTextRes: Int, explanTextRes: Int, inputLabelRes: Int, inputHintRes: Int, nextButtonRes: Int): SingleTextInputFragment {
            val fragment = SingleTextInputFragment()
            val args = Bundle()
            TextInputFragment.addStandardArgs(args, headerTextRes, explanTextRes)
            args.putInt(INPUT_LABEL_RES, inputLabelRes)
            args.putInt(INPUT_HINT_RES, inputHintRes)
            args.putInt(NEXT_BUTTON_RES, nextButtonRes)
            fragment.arguments = args
            return fragment
        }
    }
}